### PR TITLE
add feature flag to disable smp support

### PR DESCRIPTION
**Pull request by stlankes**
- to accelerate the boot time, add the option to disable SMP support

### CODE REVIEW

**Review by jbreitbart — CHANGES_REQUESTED** [review]
I still need to look at the details, but thinking about code that is not used if a feature that disables SMP support is not set makes my head hurt. 😉 Can we name the feature smp and turn it on by default? Oder does that have any downsides?

**Comment by stlankes** [comment]
Hehe, I agree

**Comment by stlankes** [comment]
bors try

**Comment by stlankes** [comment]
bors try

**Comment by stlankes** [comment]
bors try

**Comment by stlankes** [comment]
@jbreitbart Do you like the patch?

File: src/scheduler/mod.rs [hunk 1]
```diff
@@ -183,16 +197,25 @@ impl PerCoreScheduler {
...
+			#[cfg(feature = "smp")]
 			if core_id != core_scheduler().core_id {
 				input_locked.new_tasks.push_back(clone_task);
 				true
 			} else {
 				core_scheduler().ready_queue.push(clone_task);
 				false
 			}
+			#[cfg(not(feature = "smp"))]
+			if core_id == core_scheduler().core_id {
```
Comment (jbreitbart) [inline src/scheduler/mod.rs:213]: Is the core_id a constant when smp is disabled?

**Review by jbreitbart — CHANGES_REQUESTED** [review]
I looked at your changes, but I am not sure if there are no additional parts of the code that could be optimized for when SMP is disabled. The code looks mostly fine, but there are no changes to the CI. You should test the code with SMP disabled if you want to be sure it is functional. Have you run any benchmarks? Do you see any benefits from disabling SMP?

File: src/scheduler/mod.rs (hunk 1 above)
Comment (stlankes) [reply src/scheduler/mod.rs:213]: core_id is constant.

File: src/arch/x86_64/kernel/mod.rs [hunk 2]
```diff
@@ -354,12 +360,13 @@ pub fn boot_processor_init() {
...
 #[cfg(target_os = "hermit")]
 pub fn boot_application_processors() {
+	#[cfg(feature = "smp")]
 	apic::boot_application_processors();
 	apic::print_information();
 }
 
 /// Application Processor initialization
-#[cfg(target_os = "hermit")]
+#[cfg(all(target_os = "hermit", feature = "smp"))]
```
[the remark this answers is not in the document]
Comment (stlankes) [reply src/arch/x86_64/kernel/mod.rs:369]: The `target_os` is important, if we use `cargo test` for some basic tests. These tests are running on the OS, where we build the tests.  For instance, basic tests for handling the waiting queues are possible. But, every hardware related tests are not possible. In this case, we initialize the processor, which is only possible, if we boot the system.

**Comment by stlankes** [comment]
In case of our `demo`, the thread creation is a little bit faster. Also the matrix multiplication and the pi calculation. But here is the difference small and more measurement inaccuracy. The binary is 130KB smaller.

In case of `hello_wolrd`, the binary is 10 KB smaller. The runtime is also faster. But again, the difference is small and more measurement inaccuracy. 

For these rests, I use uhyve. In Qemu I would expect that the boot time for the version without SMP support is faster because we don't have to parse the multi-processor tables.

For some tests which GitHub Actions, I need also some changes in `hermitcode\rusty-hemit`. I will add these tests, after the acceptance of this PR.

**Comment by stlankes** [comment]
bors try

File: src/arch/x86_64/kernel/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jbreitbart) [reply src/arch/x86_64/kernel/mod.rs:369]: I see, thanks for the explanation. That also means it was not necessary here and removing it is fine?

https://github.com/hermitcore/libhermit-rs/pull/142/files/81b778a603299f5a09ea40c9304bc2d5aa39ec72#diff-499078c6ea49b14787bc425331ad2b552b919d74ac6add023d7c16e261a3af37L12
https://github.com/hermitcore/libhermit-rs/pull/142/files/81b778a603299f5a09ea40c9304bc2d5aa39ec72#diff-499078c6ea49b14787bc425331ad2b552b919d74ac6add023d7c16e261a3af37L535
https://github.com/hermitcore/libhermit-rs/pull/142/files/81b778a603299f5a09ea40c9304bc2d5aa39ec72#diff-ba2106a23e78160eb33cf1549af4d3328dfaa2084d11c60fd62141f5e4395812L43

File: src/scheduler/mod.rs (hunk 1 above)
Comment (jbreitbart) [reply src/scheduler/mod.rs:213]: Can you turn `core_scheduler().core_id` into a compile time constant?

File: src/arch/x86_64/kernel/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (stlankes) [reply src/arch/x86_64/kernel/mod.rs:369]: you are right, I fixed it

File: src/arch/x86_64/kernel/mod.rs [hunk 3]
```diff
@@ -429,6 +436,13 @@ unsafe fn pre_init(boot_info: &'static mut BootInfo) -> ! {
 	if boot_info.cpu_online == 0 {
 		crate::boot_processor_main()
 	} else {
-		crate::application_processor_main()
+		#[cfg(feature = "smp")]
+		crate::application_processor_main();
+		#[cfg(not(feature = "smp"))]
+		error!("SMP support deactivated");
```
[the remark this answers is not in the document]
Comment (stlankes) [reply src/arch/x86_64/kernel/mod.rs]: Ok, I fixed it

File: src/scheduler/mod.rs (hunk 1 above)
Comment (stlankes) [reply src/scheduler/mod.rs]: No, but I am able to remove it.

**Review by jbreitbart — APPROVED** [review]
bors r+

**Pull request merged**